org.eclipse.xtext.common.types.ui/src/org/eclipse/xtext/common/types/ui/trace/TraceForTypeRootProvider.java

```diff
@@ -180,13 +181,11 @@ private IEclipseTrace createTraceToSource(final ITypeRoot derivedJavaType) {
 	}
 
 	public IEclipseTrace getTraceToSource(final ICompilationUnit javaFile) {
-		try {
-			IResource resource = javaFile.getUnderlyingResource();
-			if (resource instanceof IStorage)
-				return traceForStorageProvider.getTraceToSource((IStorage) resource);
-		} catch (JavaModelException e) {
-			log.error("Error finding trace to source", e);
-		}
+		IWorkspaceRoot workspace = ResourcesPlugin.getWorkspace().getRoot();
```

Comment:
Since we do already have a resource at hand (within the javaFile), I doubt that the lookup from scratch does a better job.
Would it make sense to check the JavaModelException for e.isDoesNotExist and skip logging the error in these cases?

Reply:
Hi Sebastian, When you say better, do you mean Performance or the correctness of the fix provided ? I did some File Search tests and looks like it works fine. Do you see any problems here or you prefer guarding this case and skipping the log ?

Reply:
well to avoid the javamodelexception we would check the compilation unit for isOpen and call the getUnderlying resource only when it is. this pr targets to get it running even if the file is not in a source folder

Reply:
I see. In that case I'd move the proposed solution into the catch block.

Reply:
And what to catch for?

Reply:
sth like

```
	public IEclipseTrace getTraceToSource(final ICompilationUnit javaFile) {
		IResource resource = null;
		try {
			resource = javaFile.getUnderlyingResource();
		} catch (JavaModelException e) {
			IWorkspaceRoot workspace = ResourcesPlugin.getWorkspace().getRoot();
			resource = workspace.findMember(javaFile.getPath());
		}
		if (resource instanceof IStorage)
			return traceForStorageProvider.getTraceToSource((IStorage) resource);
		return null;
	}
```
?

Reply:
Yes, something along these lines looks good to me.

Reply:
Minor: The workspace can be injected.

Reply:
would need an optional=true + a null check

Reply:
I have fixed the code according to the review comments.